packages/http-proxy-agent/src/index.ts

```diff
@@ -107,7 +110,21 @@ export class HttpProxyAgent<Uri extends string> extends Agent {
 				this.keepAlive ? 'Keep-Alive' : 'close'
 			);
 		}
+	}
+
+	addRequest(req: InternalClientRequest, opts: AgentConnectOpts): void {
+		this.setRequestProps(req, opts);
+		// @ts-expect-error `addRequest()` isn't defined in `@types/node`
+		super.addRequest(req, opts);
+	}
 
+	async connect(
+		req: InternalClientRequest,
+		opts: AgentConnectOpts
+	): Promise<net.Socket> {
+		if (this.setRequestPropsOnConnect) {
+			this.setRequestProps(req, opts);
+		}
 		// Create a socket connection to the proxy server.
 		let socket: net.Socket;
```

Comment:
This is the primary change to fix the problem. I noticed that when `keepAlive` is enabled, the call stack for the first request is [`Agent.addRequest()`](https://github.com/nodejs/node/blob/2dd6d76c898c930d0f7d9386bde4b6dec37b7a49/lib/_http_client.js#L337) -> [`Agent.createSocket()`](https://github.com/nodejs/node/blob/2dd6d76c898c930d0f7d9386bde4b6dec37b7a49/lib/_http_agent.js#L289-L294) -> [`Agent.connect()`](https://github.com/TooTallNate/proxy-agents/blob/2f835a41f265280192b82556db80ccbe67140753/packages/agent-base/src/index.ts#L61), and all subsequent requests only call `Agent.addRequest()`.

* Before: `HttpProxyAgent.connect()` rewrites the path and add headers for only the first request
* After:
  * In "http-proxy-agent": `HttpProxyAgent.addRequest()` will rewrite the path and add headers for all requests
  * In "proxy-agent" and "pac-proxy-agent": we need to use the `setRequestPropsOnConnect` option to cause `HttpProxyAgent.connect()` to rewrite the path and add headers for the _first_ request, and then we have the right context to rewrite the path and add headers for all _subsequent_ requests.

It's not ideal to have the second code path with the `setRequestPropsOnConnect` option, but because of the way that "proxy-agent" and "pac-proxy-agent" abstract out the usage of "http-proxy-agent", I couldn't see a cleaner way to do this.

Reply:
I tweaked this in 2ddd92f2428d34c4e943e92235446ecf5ef8f5fb in order to descope this PR and maintain full backwards compatibility with ProxyAgent/PacProxyAgent, see https://github.com/TooTallNate/proxy-agents/pull/169#discussion_r1194578913